#635 StartsRequestTest sometimes fails

maurezen opened the pull request:
Added proxy option to the gpg call. 
Proxy settings are obtained via system http proxy properties.

Note to reviewer:
Somehow the changes from #701 (pull request #713) are shown in this pull request, though #713 is already merged into master.

alex-palevsky commented:
Thanks for the pull request, let me find a reviewer..

alex-palevsky commented:
@yegor256 review this pull request plz

yegor256 commented on src/main/java/com/rultor/agents/req/Decrypt.java:
```diff
@@ -58,11 +59,28 @@
+     */
+    Decrypt(final Profile prof, final String host, final int port) {
+        this.profile = prof;
+        this.proxy = this.createProxyString(host, port);
```
there should be only one "working" ctor, all others should call it, see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#constructors

yegor256 commented on src/main/java/com/rultor/agents/req/Decrypt.java:
```diff
@@ -115,4 +134,18 @@
+            "http-proxy=",
+            host,
+            ":",
+            port
```
can't you put them all in one line? looks like there is enough space

yegor256 commented on src/main/java/com/rultor/agents/req/Decrypt.java:
```diff
@@ -115,4 +134,18 @@
+     * @param port Port.
+     * @return Command string.
+     */
+    private String createProxyString(final String host, final int port) {
```
see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#method-names

yegor256 commented on src/main/java/com/rultor/agents/req/StartsRequest.java:
```diff
@@ -173,6 +190,42 @@ public boolean apply(final String input) {
+     * Obtain proxy settings and create a Decrypt instance.
+     * @return Decrypt instance.
+     */
+    private Decrypt createDecrypt() {
```
see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#method-names

yegor256 commented on src/main/java/com/rultor/agents/req/StartsRequest.java:
```diff
@@ -173,6 +190,42 @@ public boolean apply(final String input) {
+            if (StringUtils.isNotBlank(port)) {
+                try {
+                    portValue = Integer.parseInt(port);
+                } catch (final NumberFormatException exception) {
```
what it's for? why re-throwing? it's an un-checked exception.

yegor256 commented:
@maurezen a few comments above

maurezen commented:
@yegor256 re-throwing is for message composition, which should imrove log readability. Is entirely optional, though. Do you confirm that this is not needed?

I can't see where Decrypt#createProxyString doesn't conform to the linked method naming guidelines. It doesn't start with 'get' and its name reflects what it does. If it is a violation of the guideline, please do suggest an alternative name or at least specify explicitly what principle is violated. In the meantime, added a javadoc link to Decrypt#proxy.

Extracted obtaining proxy settings from StartsRequest#createDecrypt so it stops violating the naming convention.

Compacted the multiline method call to a single line.

yegor256 commented:
Here is what you do:

```
private Decrypt createDecrypt() {
```

Here is what the article (http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#method-names) says: _If a method returns something, then its name should explain what it; if it returns void, then its name should explain what it does_

So, your method should be named `decrypt()`. Make sense?

maurezen commented:
Maybe its just my English, but no, it doesn't. When I first see "decrypt" I think of a verb, not of a noun, hence the disambiguation via *create* prefix. I'll rename it if you insist, but honestly I don't think that is a good idea, since the longer name is not that long.

yegor256 commented:
@maurezen call it "decryptor()" or something like that, but don't start it with a verb if the method returns something. It's conceptually wrong. A method should either update or read. Not both at the same time.

maurezen commented:
@yegor256 renamed it to #decryptor().

How should one name a factory method in this paradigm?

yegor256 commented:
@maurezen depends on what this factory method returns. If it returns a user, for example, name it `user(int)`

yegor256 commented on src/main/java/com/rultor/agents/req/Decrypt.java:
```diff
@@ -58,11 +67,33 @@
+     * @param settings Proxy settings container
      */
-    Decrypt(final Profile prof) {
+    public Decrypt(final Profile prof, final Pair<String, Integer> settings) {
```
this `public` modifier is redundant, since the class is visible only inside the package

yegor256 replied:
why `Pair`? why can't you pass `String` and `int`?

yegor256 commented on src/main/java/com/rultor/agents/req/Decrypt.java:
```diff
@@ -115,4 +147,14 @@
+     * @return Command string.
+     * @see #proxy
+     */
+    private static String createProxyString(final String host, final int port) {
```
same here, don't start with a verb (`create`) if the method is NOT updating the object

yegor256 commented:
@maurezen a few more comments above

maurezen commented:
@yegor256  I want to get proxy settings in a separate method and a method can only return one value, hence the Pair.

Fixed modifiers.

yegor256 commented:
@maurezen this `Pair` thing looks very wrong. instead of solving your design issues internally, you're putting the problem on the shoulders of the class users. imagine, I will use your class in a few months. and I will see this `Pair` as an argument. I'll be very surprised... at least :) let's refactor this. it's a bad design.

maurezen commented:
@yegor256 I can see the "least wtf principle" (or YAGNI) as one of the design principles, but not as the sole one, so let's discuss why it is a bad design. 

What I want is to encapsulate the proxy settings in one entity, seeing as proxy host is useless without a proxy port and vice versa, and to decouple data acquisition from Decrypt instance creation. Actually I could write a holder for this pair of values, but 
 - It is introducing a separate entity while we can use an existing one perfectly fine, so 
 - Dependency validator fails as this would 8th data abstraction on which Decrypt is dependant. This could probably be circumvented by uncoupling actual data storage and data access interface, but in the light of the first reason this looks like overengineering squared.

So, is the proxy settings encapsulation really wrong? If yes, why is that?

maurezen commented:
@yegor256 it's been 36 hours since my commits, and I've seen you answer me at least twice in other threads. Could you please continue the review process?

yegor256 commented:
@maurezen here is how I would do this:

```
Decrypt(Profile prof, String opts) {
  this.profile = prof;
  this.options = opts;
}
Decrypt(Profile prof) {
  this(prof, "");
}
```

Then, in unit tests:

```
new Decrypt(profile, "http-proxy=host:port");
```

In other words, it's not important for the `Decrypt` class what gpg options we're sending. All it needs to know is that we want to add some options to the gpg call. What do you think?

maurezen commented:
@yegor256 it doesn't look right, either. Suddenly, we have
 - an ability to pass arbitrary option to the call, which is potential point of failure
 - additional pieces of code that are aware of gpg and its option format, which is another point of failure (imagine we want to switch to another tool. We now need to change Decrypt implementation AND all of the constructor calls. Shiny!)

That's why we want to interpret the proxy settings inside Decrypt.

yegor256 commented:
@maurezen fair enough. So, how about this:

```
Decrypt(Profile prof, String host, int port) {
  this.profile = prof;
  if (port == 0) {
    this.options = "";
  } else {
    this.options = String.format("http-proxy=%s:%d", host, port);
  }
}
Decrypt(Profile prof) {
  this(prof, "", 0);
}
```

Then, you get rid of `obtainProxySettings()` and put its code into `decryptor()`.

Besides, try to avoid `StringUtils` usages that often. We don't need their `isBlank` or similar methods because we never work with NULL. So, we don't need to be too protective. Besides, `StringUtils.EMPTY` is also a redundancy. Just `""` is shorter and clear enough.

maurezen commented:
decryptor() becomes a method doing two distinctly different things at this point, but I could live with that, since we still get the proxy settings in one place.

Do you want StringUtils removed from this pull request, or is that a comment for future consideration?

yegor256 commented:
@maurezen I would suggest to remove it from this PR as much as possible.

maurezen commented:
@yegor256 done, please take a look.

yegor256 commented on src/main/java/com/rultor/agents/req/Decrypt.java:
```diff
@@ -115,4 +143,14 @@
+     * @return Command string.
+     * @see #proxy
+     */
+    private static String proxyString(final String host, final int port) {
```
why so many lines of code if we can just do `String.format()` right in the constructor? this method looks like over-engineering to me.

yegor256 commented on src/main/java/com/rultor/agents/req/StartsRequest.java:
```diff
@@ -173,6 +187,36 @@ public boolean apply(final String input) {
+            if (!port.isEmpty()) {
+                try {
+                    portValue = Integer.parseInt(port);
+                } catch (final NumberFormatException exception) {
```
what is it for? it's an unchecked exception, why do you catch it here?

yegor256 commented on src/test/java/com/rultor/agents/req/DecryptTest.java:
```diff
@@ -76,6 +77,11 @@
     /**
+     * Test profile name.
+     */
+    private static final String TEST_PROFILE_NAME = "test/test";
```
this is a bad idea, see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#redundant-constants

yegor256 commented on src/test/java/com/rultor/agents/req/DecryptTest.java:
```diff
@@ -133,4 +134,41 @@ public void decryptsAssets() throws Exception {
+            "http://someserver.com",
+            8080
+        );
+        final Iterable<String> commands = decrypt.commands();
```
it's a one time variable, see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#one-time-variables

yegor256 commented:
@maurezen a few more comments above

maurezen commented:
@yegor256 
>String-format
Agreed, will be fixed.

>Re-throw
Answered this one already: re-throwing is for message composition, which should imrove log readability. Is entirely optional, though. Do you confirm that this is not needed?

>constant
It is used in more than one place, though. And as soon as someone needs to test something else in Decrypt, it will be used more. Do you confirm that I should inline it in multiple places?

>one-time variable
It reads better than a multiline constructor immediately followed by a method call. Actually, both variables in DecryptTest#testHttpProxyHandling are one-time and could be inlined, but do we really need that? Readability will degrade, while the performance gain is dubious (and while I cannot find anything definite, either compiler or JIT will inline these calls anyway).

yegor256 commented:
I agree about re-throw, let's leave it.

The thing with constant is that it makes an impression to the reader that there is some link between test methods. There is something in common. Why there is nothing. That's why it's better to use unique string literals in every test method.

Yes, one-time variables are bad and should be inlined as much as possible. They are simply adding noise to the code

maurezen commented:
@yegor256 (not so) fun fact: as soon as I inline something in that method, magic number code style check starts to fail. Will have to introduce a constant.

yegor256 commented:
@maurezen yes, you have to use different string literals in each method

maurezen commented:
@yegor256 ok, done.

yegor256 commented on src/main/java/com/rultor/agents/req/StartsRequest.java:
```diff
@@ -173,6 +187,36 @@ public boolean apply(final String input) {
+                        Joiner.on(' ').join(
+                            "Can't parse proxy port",
+                            port
+                        ),
```
again, why four lines of code when you may do it in one: `String.format("can't parse proxy port %d", port)`. check this http://www.yegor256.com/2014/06/19/avoid-string-concatenation.html

yegor256 commented:
@maurezen to be honest, this decryptor method looks over-complex. I would do it like this:

```
return new Decrypt(
  this.profile, 
  System.getProperty(Decrypt.HOST_KEY, ""), 
  Integer.parseInt(System.getProperty(Decrypt.PORT_KEY, "0"))
);
```

Don't you think?

maurezen commented:
@yegor256 done.

yegor256 commented:
@rultor try to merge pls

rultor commented:
> @rultor try to merge pls

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/1538-69476424)

rultor commented:
> @rultor try to merge pls

@yegor256 Oops, I failed. You can see the full log [here](http://www.rultor.com/t/1538-69476424) (spent 8min)

maurezen commented:
@yegor256 the automatic merge failed. what do we do now?

yegor256 commented:
@rultor try to merge again

rultor commented:
> @rultor try to merge again

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/1538-69511748)

rultor commented:
> @rultor try to merge again

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/1538-69511748) (took me 14min)

alex-palevsky commented:
@yegor256 I added **46 mins** to your account, many thanks for working with the project!

+46 added to your rating, at the moment it is: [+4937](http://www.netbout.com/b/35008?open=rating)

alex-palevsky commented:
@rultor deploy now pls

rultor commented:
> @rultor deploy now pls

@alex-palevsky OK, I'll try to deploy now. You can check the progress [here](http://www.rultor.com/t/1538-69916790)

rultor commented:
> @rultor deploy now pls

@alex-palevsky Done! FYI, the full log is [here](http://www.rultor.com/t/1538-69916790) (took me 8min)
